lib/register.php

```diff
@@ -82,39 +138,146 @@ function gutenberg_add_edit_links_filters() {
  *
  * @since 0.1.0
  *
- * @param  array $actions Post actions.
- * @param  array $post    Edited post.
+ * @param  array   $actions Post actions.
+ * @param  WP_Post $post    Edited post.
  *
  * @return array          Updated post actions.
  */
 function gutenberg_add_edit_links( $actions, $post ) {
-	$can_edit_post = current_user_can( 'edit_post', $post->ID );
-	$title = _draft_or_post_title( $post->ID );
-	$post_type = get_post_type( $post );
+	if ( ! gutenberg_can_edit_post( $post->ID ) ||
+			'trash' === $post->post_status ||
+			! post_type_supports( $post->post_type, 'editor' ) ||
+			! apply_filters( 'gutenberg_add_edit_link_for_post_type', true, $post->post_type, $post ) ) {
+		return $actions;
+	}
 
-	if ( $can_edit_post && 'trash' !== $post->post_status && apply_filters( 'gutenberg_add_edit_link_for_post_type', true, $post_type, $post ) ) {
-		// Build the Gutenberg edit action. See also: WP_Posts_List_Table::handle_row_actions().
-		$gutenberg_url = menu_page_url( 'gutenberg', false );
-		$gutenberg_action = sprintf(
+	if ( gutenberg_post_has_blocks( $post->ID ) ) {
+		remove_filter( 'get_edit_post_link', 'gutenberg_filter_edit_post_link', 10 );
+		add_filter( 'get_edit_post_link', 'gutenberg_filter_edit_post_link', 10, 3 );
```

Comment:
@westonruter Should we just add a default filter handler for `get_edit_post_link` to always override with the preferred editor by content?

Reply:
Isn't that what is already done here?

https://github.com/WordPress/gutenberg/blob/547d0386c83389e848d9a2c213c50ea5e9749bc9/lib/register.php#L210-L231

Reply:
The reason why the `remove_filter` and `add_filter` are done above is to facilitate obtaining the classic post editor URL, even for a post that has blocks in it:

https://github.com/WordPress/gutenberg/blob/547d0386c83389e848d9a2c213c50ea5e9749bc9/lib/register.php#L88-L91

Reply:
Yeah, seeing that now in the changes prior to 63ff18dd862a564d3ab7513a52bdbd84c551e436. I guess this might not be working correctly for the "Classic Editor" link for Gutenberg posts currently then.